src/Tools/_framework/Paths/Courses.jsx

```diff
@@ -0,0 +1,557 @@
+// import axios from 'axios';
+import {
+  Box,
+  Button,
+  Text,
+  Grid,
+  GridItem,
+  Wrap,
+  Flex,
+  DrawerFooter,
+  FormErrorMessage,
+  Input,
+  FormLabel,
+  FormControl,
+  FormHelperText,
+  Stack,
+  DrawerBody,
+  DrawerHeader,
+  DrawerCloseButton,
+  DrawerContent,
+  DrawerOverlay,
+  Drawer,
+  useDisclosure,
+  Image,
+  Popover,
+  PopoverTrigger,
+  PopoverContent,
+  PopoverArrow,
+  PopoverBody,
+  SimpleGrid,
+  AlertDialog,
+  AlertDialogOverlay,
+  AlertDialogContent,
+  AlertDialogHeader,
+  AlertDialogBody,
+  AlertDialogFooter,
+} from "@chakra-ui/react";
+
+import React, { useEffect, useRef, useState } from "react";
+import { useLoaderData, useFetcher } from "react-router-dom";
+import axios from "axios";
+import { CourseCard } from "../../../_reactComponents/PanelHeaderComponents/CourseCard";
+import { driveColors, driveImages } from "../../../_reactComponents/Drive/util";
+
+export async function action({ request }) {
+  const formData = await request.formData();
+  let formObj = Object.fromEntries(formData);
+
+  if (formObj?._action == "Create New Course") {
+    let { data } = await axios.get("/api/createCourse.php");
+    if (!data.success) throw Error(data.message);
+    return true;
+  } else if (formObj?._action == "Delete") {
+    let { data } = await axios.post("/api/deleteCourse.php", {
+      courseId: formObj.courseId,
+    });
+    if (!data.success) throw Error(data.message);
+    return true;
+  } else if (formObj?._action == "Duplicate") {
+    let { data } = await axios.post("/api/duplicateCourse.php", {
+      courseId: formObj.courseId,
+      dateDifference: formObj.dateDifference,
+      newLabel: formObj.newLabel,
+    });
+    if (!data.success) throw Error(data.message);
+    return true;
+  } else if (formObj?._action == "Rename") {
+    let { data } = await axios.post("/api/modifyCourse.php", {
+      courseId: formObj.courseId,
+      label: formObj.newLabel,
+    });
+    if (!data.success) throw Error(data.message);
```

Comment:
This doesn't work when the API returns error codes, the call to await an axios.post() will itself throw an error. Currently there isn't much validation on these APIs that is easy to trigger from the front-end, but for the sake of fault tolerant front-end I think we want to do this better.

This is what happens when I modify the modifyCourse api to unconditionally throw an exception, and thus make the script return a 400 error code with the try/catch pattern. This also shows why error boundaries aren't really the only error handling we want to have, because they kill any part of the page that is under the current Outlet.


https://github.com/Doenet/DoenetTools/assets/365366/97ad94bd-bc26-4d34-b08e-74fd00ae5b5f

Reply:
Agreed. Error handling needs work for sure.  
https://github.com/Doenet/DoenetTools/issues/2273

Reply:
Added Error Handling